src/providers/log-action.provider.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: loopback4-example-log-extension
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {inject, Provider, Constructor, Getter} from '@loopback/context';
+import {CoreBindings} from '@loopback/core';
+import {OperationArgs, ParsedRequest} from '@loopback/rest';
+import {getLogMetadata} from '../decorators/log.decorator';
+import {EXAMPLE_LOG_BINDINGS, LOG_LEVEL} from '../keys';
+import {LogFn, Time, TimerFn} from '../types';
+
+export class LogActionProvider implements Provider<LogFn> {
+  constructor(
+    @inject.getter(CoreBindings.CONTROLLER_CLASS)
+    private readonly getController: Getter<Constructor<{}>>,
+    @inject.getter(CoreBindings.CONTROLLER_METHOD_NAME)
+    private readonly getMethod: Getter<string>,
+    @inject(EXAMPLE_LOG_BINDINGS.APP_LOG_LEVEL)
+    private readonly logLevel: number,
+    @inject(EXAMPLE_LOG_BINDINGS.TIMER) public timer: TimerFn,
+  ) {}
+
+  value() {
+    const fn = <LogFn>(async (
+      req: ParsedRequest,
+      args: OperationArgs,
+      // tslint:disable-next-line:no-any
+      result: any,
+      start?: [number, number],
+    ) => {
+      const controllerClass = await this.getController();
```

Comment:
I was thinking about this pattern recently. When something wrong happens inside a sequence action implemented as an anonymous function, the stack trace will contain an empty function name that's rather unhelpful to diagnose the problem.

I am proposing to move the meat of the sequence action into a named function in the provider class, for example `executeAction`.

```ts
export class LogActionProvider implements Provider<LogFn> {
  // ...
  value() {
    // notice we don't need "async" keyword here
    const fn = <LogFn>(
       // ...args...
    ) => {
      return this.executeAction(/*forward args*/);
    };
  }

  private async executeAction(
      req: ParsedRequest,
      args: OperationArgs,
      // tslint:disable-next-line:no-any
      result: any,
      start?: [number, number],
  ) {
    const controllerClass = await this.getController();
    // etc.
  }
}
```

This way, the stack trace will show `LogActionProvider.executeAction`.